mqtt_as/mqtt_as.py

```diff
@@ -322,26 +334,34 @@ def close(self):
     async def publish(self, topic, msg, retain, qos):
         self.pid = newpid(self.pid)  # Why did I only update self.pid if qos == 1 ??
         pid = self.pid  # Keep local in case self.pid is updated
+        self.dprint("publish", topic, "pid", pid)
         async with self.lock:
-            await self._publish(topic, msg, retain, qos, 0, pid)  # ._publish adapted to use passed pid
+            self.dprint("publish lock", topic, pid)
+            await self._publish(topic, msg, retain, qos, 0,
+                                pid)  # ._publish adapted to use passed pid
         if qos == 0:
             return
 
         count = 0
         while 1:  # Await PUBACK, republish on timeout
             t = ticks_ms()
+            self.dprint("awaiting pid", pid)
             while pid not in self.rcv_pids:  # local copy
                 await asyncio.sleep_ms(200)
                 if self._timeout(t) or not self.isconnected():
```

Comment:
check again if pids in rcv_pids because in my test 3 pids were already received when the connection closed, so those 3 publish bailed here and got republished although they already had their pid received.